JENKINS-25209: Add "reboot slave after build" for a slave template

arcivanov opened the pull request:
Adds slave template option to reboot a slave after a build
Once reboot commences the slave is monitored, reconnects are started and
the slave is reenabled once reconnect succeeds.
Fixes NPEs in tag handling
Make constructors more tolerant of null-strings.
Fix EC2ComputerLauncher switch state handling and breaks-to-label

arcivanov commented:
retest this please

francisu commented:
Thanks Arcadiy, I'm having trouble testing this as there seems to be some problem with Eclipse when using the more recent Jenkins libraries. Please give me some more time on this. I'm not going to take anything else though until I have finished with your pull request.

arcivanov commented:
Thanks, Francis. There is something weird going on on Eclipse with older libraries indeed. I've had to modify pom to use 1.565.3 while testing locally and it worked fine. I would then change the pom.xml back to the older version of Jenkins libraries before compiling from the command line/CI and checking in.

francisu commented:
Hi Arcadiy, I spent a little time with this. I changed the version in the pom to use 1.565.3 and I had to remove the "@DataBoundConstructor" from EC2AbstractSlave to get it to compile. And then when I ran the tests, a couple of them failed: http://pastebin.com/Xqu3NNyN.  With the original pom version, 35 of the tests failed (just running "mvn clean test" I spent some time trying to work out the problems with the test that I think they are in the handling of passwords, probably using the Secret class, but I ran out of time I could spend to try and come up with a fix. If you are unable to reproduce the original test failures, let me know and I can give you the output. Can you look at this some more? 
Thanks,
Francis

buildhive commented:
[Jenkins » ec2-plugin #105](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/105/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

arcivanov commented:
This is the reason why the tests fail:

WARNING: Caught exception evaluating: h.getPasswordValue(attrs.value ?: instance[attrs.field]) in /configure. Reason: java.lang.SecurityException: attempted to render plaintext ‘Rc9PPykm5L1LpETfW45sWXV9A/5+rNILB2tMxC9VPt8=’ in password field; use a getter of type Secret instead

The problem is that 	at hudson.Functions.getPasswordValue(Functions.java:1682) we now force unit test failures for security reasons.

buildhive commented:
[Jenkins » ec2-plugin #106](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/106/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Jenkins » ec2-plugin #117](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/117/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Jenkins » ec2-plugin #119](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/119/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Jenkins » ec2-plugin #150](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/150/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Jenkins » ec2-plugin #151](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/151/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Jenkins » ec2-plugin #153](https://buildhive.cloudbees.com/job/jenkinsci/job/ec2-plugin/153/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

johnnyshields commented:
Just a note, this option should not be enabled for spot instances as they cannot reboot.

johnnyshields commented:
@francisu I will pickup this feature and combine it with some others I have planned. We can leave this PR open for now.

arcivanov commented:
@johnnyshields was this merged in one form or the other?

johnnyshields commented:
Not yet, if you want to help with it you're welcome to.

arcivanov commented:
@johnnyshields help with what? :smile: 

I have rebased on the latest master and this is ready for merge on its own merits. I have removed the reboot option for spot instances per above and the option will be ignored for them.

arcivanov commented:
@johnnyshields @francisu if you need "help", could you at least specify help with what?

johnnyshields commented:
@arcivanov sorry for holding up the show here. I'm fine with this PR as is from looking at the code. I assume you've tested it locally and no issue? @francisu ?

arcivanov commented:
@johnnyshields Sorry for late response. Yes, I confirm the branch in the current state works in our production. Thanks!
